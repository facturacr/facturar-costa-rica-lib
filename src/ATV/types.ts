export type Mode = 'prod' | 'stg' | 'custom';

export type ATVOptions = {
  defaultOverwrites?: {
    tokenServiceUrl: string;
  };
}

export type TokenServiceProps = {
  serviceUrl: string;
  clientId: string;
}

export type SendResponse = {
  status: number;
  location: string;
  errorCause?: string | null;
}

export type SendConfirmationInput = {
  url: string;
  headers: {
    Authorization: string;
    'Content-Type': string;
  };
}

export type ConfirmationMessage = {
  clave: string;
  message: string;
  details: string;
  emitterFullName: string;
  emitterIdentifierId: string;
  receiverIdentifierId: string;
  taxesTotalAmount: string;
  totalInvoiceAmount: string;
}