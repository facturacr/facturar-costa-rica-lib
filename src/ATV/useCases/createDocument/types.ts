import { InvoiceDocumentContainer } from '@src/types/facturaInterfaces'
import { Method } from 'axios'

export type DocumentTypes = 'FE'

type CommandData = {
  clave: string;
  fecha: string;
  emisor: {
    tipoIdentificacion: string;
    numeroIdentificacion: string;
  };
  receptor: {
    tipoIdentificacion: string;
    numeroIdentificacion: string;
  };
  comprobanteXml: string;
  consecutivoReceptor?: string;
};

export type Command = {
  url: string;
  method: Method;
  data: CommandData;
  headers: {
    Authorization: string;
    'Content-Type': string;
  };
}

export type CreateAndSendDocumentResponse = {
  command: Command;
  extraData: {
    xml: string;
    document: InvoiceDocumentContainer;
  };
}
