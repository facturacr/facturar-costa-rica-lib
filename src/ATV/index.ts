
import axios from 'axios'
import { GetToken } from '../services/getToken/GetToken'
import { GetTokenDto, GetTokenResponse } from '../services/getToken/types'
import { CreateDocumentCommand, CreateDocumentInput } from './useCases/sendDocument/SendDocument'
import { Command, CreateAndSendDocumentResponse } from './useCases/sendDocument/types'
import * as parser from 'fast-xml-parser'
import qs from 'querystring'
import { ConfirmationMessageRaw } from '@src/types/facturaInterfaces'

export type Mode = 'prod' | 'stg' | 'custom';

export type ATVOptions = {
  defaultOverwrites?: {
    tokenServiceUrl: string;
  };
}

export type TokenServiceProps = {
  serviceUrl: string;
  clientId: string;
}

export type SendResponse = {
  status: number;
  location: string;
  errorCause?: string | null;
}

type SendConfirmationInput = {
  url: string;
  headers: {
    Authorization: string;
    'Content-Type': string;
  };
}

type ConfirmationMessage = {
  clave: string;
  message: string;
  details: string;
  emitterFullName: string;
  emitterIdentifierId: string;
  receiverIdentifierId: string;
  taxesTotalAmount: string;
  totalInvoiceAmount: string;
}

export class ATV {
  public readonly options: ATVOptions
  constructor(options?: ATVOptions, public readonly mode: Mode = 'prod') {
    this.options = options
  }

  public getToken(params: GetTokenDto): Promise<GetTokenResponse> {
    const tokenService = new GetToken(this)
    return tokenService.execute(params)
  }

  public createDocumentCommand(input: CreateDocumentInput): Promise<CreateAndSendDocumentResponse> {
    const sendDocument = new CreateDocumentCommand(this)
    return sendDocument.execute(input)
  }

  public async sendDocument(input: Command): Promise<SendResponse> {
    try {
      const response = await axios(input)
      return {
        status: response.status,
        location: response.headers.location,
        errorCause: null
      }
    } catch (error) {
      const response = error.response || {}
      const header = response.headers || {}
      return {
        status: error.response.status || 500,
        errorCause: header['x-error-cause'],
        location: null
      }
    }
  }

  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  public async sendConfirmation(input: SendConfirmationInput) {
    try {
      const response = await axios(input)
      const xmlResponse = response.data['respuesta-xml']
      if (!xmlResponse) {
        const state = response.data['ind-estado']
        return {
          status: response.status,
          state,
          errorCause: null
        }
      }
      const xmlString = this.decodeBase64(xmlResponse)

      return {
        status: response.status,
        confirmation: this.parseConfirmationMessage(xmlString),
        xml: xmlString,
        errorCause: null
      }
    } catch (error) {
      const response = error.response || {}
      return {
        status: response.status || 500,
        errorCause: response.statusText
      }
    }
  }

  private parseConfirmationMessage(xml: string): ConfirmationMessage {
    const parsedXml: ConfirmationMessageRaw = parser.parse(xml, {
      ignoreAttributes: false,
      tagValueProcessor: a => qs.unescape(a.replace(/(\r\n|\n|\r)/gm, '')),
      ignoreNameSpace: false,
      parseNodeValue: false
    })
    const msj = parsedXml.MensajeHacienda
    return {
      clave: msj.Clave,
      message: msj.Mensaje,
      details: msj.DetalleMensaje,
      emitterFullName: msj.NombreEmisor,
      emitterIdentifierId: msj.NumeroCedulaEmisor,
      receiverIdentifierId: msj.NumeroCedulaReceptor,
      taxesTotalAmount: msj.MontoTotalImpuesto,
      totalInvoiceAmount: msj.TotalFactura
    }
  }

  private decodeBase64(encodedStr: string): string {
    const buff = Buffer.from(encodedStr, 'base64')
    return buff.toString('utf-8')
  }
}
