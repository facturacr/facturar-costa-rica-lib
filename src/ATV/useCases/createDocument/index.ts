import { Person } from '@src/ATV/core/Person'
import { OrderLine } from '@src/ATV/core/OrderLine'
import { FullConsecutive } from '@src/ATV/core/FullConsecutive'
import { Document } from '@src/ATV/core/Document'
import { Clave } from '@src/ATV/core/Clave'
import { mapDocumentToAtvFormat as mapBillToAtvFormat } from '@src/ATV/mappers/billDocToAtv'
import { genXML } from '@src/lib/genXML'
import { CreateAndSendDocumentResponse, CreateDocumentInput, Command } from './types'
import { ATV } from '@src/ATV'
import { DocumentType } from '@src/ATV/core/DocumentType'
export type { CreateDocumentInput as CreateBillInput } from './types'

const options: { [key: string]: { serviceUrl: string}} = {
  prod: {
    serviceUrl: 'https://api.comprobanteselectronicos.go.cr/v1/recepcion'
  },
  stg: {
    serviceUrl: 'https://api-sandbox.comprobanteselectronicos.go.cr/recepcion/v1/recepcion'
  }
}

export class CreateDocumentCommand {
  private readonly serviceUrl: string;

  constructor(scope: ATV) {
    this.serviceUrl = options[scope.mode].serviceUrl
  }

  public async execute(dto: CreateDocumentInput): Promise<CreateAndSendDocumentResponse> {
    const documentName = dto.document.documentName || 'FacturaElectronica'
    const document = this.createDocument(dto.document)
    const atvDocument = mapBillToAtvFormat(documentName, document)
    const xml = await genXML(documentName, atvDocument, dto.signatureOptions)
    const command = await this.createDocumentCommand(document, xml, dto.token)
    return {
      command,
      extraData: {
        xml
      }
    }
  }

  private async createDocumentCommand(document: Document, xml: string, token: string): Promise<Command> {
    return {
      url: this.serviceUrl,
      method: 'post',
      data: {
        clave: document.clave,
        fecha: document.issueDate.toISOString(),
        emisor: {
          tipoIdentificacion: document.emitter.identifierType,
          numeroIdentificacion: document.emitter.identifierId
        },
        receptor: {
          tipoIdentificacion: document.receiver.identifierType,
          numeroIdentificacion: document.receiver.identifierId
        },
        comprobanteXml: this.encodeXML(xml)
      },
      headers: {
        Authorization: 'bearer ' + token,
        'Content-Type': 'application/json'
      }
    }
  }

  private createDocument(document: CreateDocumentInput['document']): Document {
    const documentType = DocumentType.create(document.documentName)
    const emitter = this.createEmitter(document.emitter)
    const receiver = this.createReceiver(document.receiver)
    const clave = this.createClave(document, documentType)
    const fullConsective = this.createFullConsecutive(document, documentType)
    const orderLines = this.createOrderLines(document)
    return Document.create({
      clave: clave,
      fullConsecutive: fullConsective,
      orderLines: orderLines,
      activityCode: document.activityCode,
      issueDate: new Date(),
      emitter: emitter,
      receiver: receiver,
      conditionSale: '01',
      paymentMethod: '01'
    })
  }

  private createEmitter(emitterInput: CreateDocumentInput['document']['emitter']): Person {
    return Person.create(emitterInput)
  }

  private createReceiver(receiverInput: CreateDocumentInput['document']['receiver']): Person {
    return Person.create(receiverInput)
  }

  private createOrderLines(dto: CreateDocumentInput['document']): OrderLine[] {
    return dto.orderLines.map((orderLine, index) => {
      return OrderLine.create({
        detail: orderLine.detail,
        unitaryPrice: orderLine.unitaryPrice,
        lineNumber: orderLine.lineNumber || (index + 1).toString(),
        code: orderLine.code,
        quantity: orderLine.quantity,
        measureUnit: orderLine.measureUnit,
        totalAmount: orderLine.totalAmount,
        tax: orderLine.tax
      })
    })
  }

  private createFullConsecutive(dto: CreateDocumentInput['document'], docType: DocumentType): FullConsecutive {
    return FullConsecutive.create({
      consecutiveIdentifier: dto.consecutiveIdentifier,
      branch: dto.branch,
      terminal: dto.terminal,
      documentType: docType.value
    })
  }

  private createClave(dto: CreateDocumentInput['document'], docType: DocumentType): Clave {
    return Clave.create({
      branch: dto.branch,
      ceSituation: dto.ceSituation,
      consecutiveIdentifier: dto.consecutiveIdentifier,
      countryCode: dto.countryCode,
      docKeyType: docType.value,
      emitterIdentifier: dto.emitter.identifier.id,
      identifierType: dto.emitter.identifier.type, // TODO add default
      securityCode: dto.securityCode,
      terminal: dto.terminal
    })
  }

  private encodeXML(xmlStr: string): string {
    const buffer = Buffer.from(xmlStr)
    return buffer.toString('base64')
  }
}
