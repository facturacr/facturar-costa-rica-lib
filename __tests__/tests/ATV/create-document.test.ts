import fs from 'fs'
import { ATV } from '@src/ATV'
import 'jest-xml-matcher'
import { createDocumentInputExample } from '@test/stubs/createDocument.data'
const fakePem = fs.readFileSync('__tests__/stubs/dummyKeys/client-identity.p12', 'binary')
const fakePassword = '1234'
const expectXml = fs.readFileSync('__tests__/stubs/commonExpectedXml.xml', 'utf-8')

describe('Create Document (Invoice)', () => {
  beforeAll(() => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date(2024, 3, 1, 0, 0, 0, 0))
  })

  it('should create document and generate a correct command', async () => {
    const atv = new ATV({}, 'stg')
    const createdDoc = await atv.createDocumentCommand({
      document: createDocumentInputExample,
      token: 'fake-token',
      signatureOptions: {
        buffer: fakePem,
        password: fakePassword
      }
    })
    expect(createdDoc.command).toMatchObject({
      url: 'https://api-sandbox.comprobanteselectronicos.go.cr/recepcion/v1/recepcion',
      method: 'post'
    })
    expect(createdDoc.command.data).toMatchObject({
      clave: '50601042400020692014200100001010000000002100000001',
      fecha: '2024-04-01T00:00:00.000Z',
      emisor: { tipoIdentificacion: createDocumentInputExample.emitter.identifier.type, numeroIdentificacion: createDocumentInputExample.emitter.identifier.id },
      receptor: { tipoIdentificacion: createDocumentInputExample.receiver.identifier.type, numeroIdentificacion: createDocumentInputExample.receiver.identifier.id }
    })
    expect(createdDoc.command.data.comprobanteXml).toBeDefined()
    expect(createdDoc.extraData.xml).toBeDefined()
  })

  it('should create document and match to snapshot', async () => {
    const atv = new ATV({}, 'stg')
    // @ts-ignore just for testing
    const createdDoc = await atv.createDocumentCommand({
      document: createDocumentInputExample,
      token: 'fake-token',
      signatureOptions: undefined
    })
    expect(createdDoc.command.data.comprobanteXml).toMatchSnapshot()
  })
  it('should create document and generate a correct xml', async () => {
    const atv = new ATV({}, 'stg')
    // @ts-ignore just for testing
    const createdDoc = await atv.createDocumentCommand({
      document: createDocumentInputExample,
      token: 'fake-token',
      signatureOptions: undefined
    })
    expect(createdDoc.extraData.xml).toEqualXML(expectXml)
  })
})
