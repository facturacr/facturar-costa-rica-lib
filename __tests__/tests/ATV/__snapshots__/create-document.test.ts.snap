// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Document (Invoice) should create document and match to snapshot 1`] = `"PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz48RmFjdHVyYUVsZWN0cm9uaWNhIHhtbG5zPSJodHRwczovL2Nkbi5jb21wcm9iYW50ZXNlbGVjdHJvbmljb3MuZ28uY3IveG1sLXNjaGVtYXMvdjQuMy9mYWN0dXJhRWxlY3Ryb25pY2EiIHhtbG5zOmRzPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwLzA5L3htbGRzaWcjIiB4bWxuczp4c2Q9Imh0dHA6Ly93d3cudzMub3JnLzIwMDEvWE1MU2NoZW1hIiB4bWxuczp4c2k9Imh0dHA6Ly93d3cudzMub3JnLzIwMDEvWE1MU2NoZW1hLWluc3RhbmNlIj48Q2xhdmU+NTA2MDEwNDI0MDAwMjA2OTIwMTQyMDAxMDAwMDEwMTAwMDAwMDAwMDIxMDAwMDAwMDE8L0NsYXZlPjxDb2RpZ29BY3RpdmlkYWQ+MDAwMDA0PC9Db2RpZ29BY3RpdmlkYWQ+PE51bWVyb0NvbnNlY3V0aXZvPjAwMTAwMDAxMDEwMDAwMDAwMDAyPC9OdW1lcm9Db25zZWN1dGl2bz48RmVjaGFFbWlzaW9uPjIwMjQtMDQtMDFUMDA6MDA6MDAuMDAwWjwvRmVjaGFFbWlzaW9uPjxFbWlzb3I+PE5vbWJyZT5FbWlzb3IgbmFtZTwvTm9tYnJlPjxJZGVudGlmaWNhY2lvbj48VGlwbz4wMTwvVGlwbz48TnVtZXJvPjIwNjkyMDE0MjwvTnVtZXJvPjwvSWRlbnRpZmljYWNpb24+PE5vbWJyZUNvbWVyY2lhbD5lbWlzb3IgY29tZXJjaWFsIG5hbWU8L05vbWJyZUNvbWVyY2lhbD48VWJpY2FjaW9uPjxQcm92aW5jaWE+MjwvUHJvdmluY2lhPjxDYW50b24+MDY8L0NhbnRvbj48RGlzdHJpdG8+MDQ8L0Rpc3RyaXRvPjxCYXJyaW8+MDY8L0JhcnJpbz48T3RyYXNTZW5hcz5kZXRhaWxzPC9PdHJhc1NlbmFzPjwvVWJpY2FjaW9uPjxUZWxlZm9ubz48Q29kaWdvUGFpcz41MDY8L0NvZGlnb1BhaXM+PE51bVRlbGVmb25vPjEyMzQxMjM0PC9OdW1UZWxlZm9ubz48L1RlbGVmb25vPjxGYXg+PENvZGlnb1BhaXM+NTA2PC9Db2RpZ29QYWlzPjxOdW1UZWxlZm9ubz4xMjM0MTIzNDwvTnVtVGVsZWZvbm8+PC9GYXg+PENvcnJlb0VsZWN0cm9uaWNvPnRlc3RAdGVzdC5jb208L0NvcnJlb0VsZWN0cm9uaWNvPjwvRW1pc29yPjxSZWNlcHRvcj48Tm9tYnJlPnJlY2VpdmVyIG5hbWUgU1JMPC9Ob21icmU+PElkZW50aWZpY2FjaW9uPjxUaXBvPjAxPC9UaXBvPjxOdW1lcm8+MjA2OTMwMTQzPC9OdW1lcm8+PC9JZGVudGlmaWNhY2lvbj48Tm9tYnJlQ29tZXJjaWFsPnJlY2VpdmVyIG5hbWU8L05vbWJyZUNvbWVyY2lhbD48VWJpY2FjaW9uPjxQcm92aW5jaWE+MjwvUHJvdmluY2lhPjxDYW50b24+MDY8L0NhbnRvbj48RGlzdHJpdG8+MDQ8L0Rpc3RyaXRvPjxCYXJyaW8+MDY8L0JhcnJpbz48T3RyYXNTZW5hcz5kZXRhaWxzPC9PdHJhc1NlbmFzPjwvVWJpY2FjaW9uPjxUZWxlZm9ubz48Q29kaWdvUGFpcz41MDY8L0NvZGlnb1BhaXM+PE51bVRlbGVmb25vPjgwODA4MDgwPC9OdW1UZWxlZm9ubz48L1RlbGVmb25vPjxGYXg+PENvZGlnb1BhaXM+NTA2PC9Db2RpZ29QYWlzPjxOdW1UZWxlZm9ubz44MDgwODA4MDwvTnVtVGVsZWZvbm8+PC9GYXg+PENvcnJlb0VsZWN0cm9uaWNvPnRlc3RAdGVzdC5jb208L0NvcnJlb0VsZWN0cm9uaWNvPjwvUmVjZXB0b3I+PENvbmRpY2lvblZlbnRhPjAxPC9Db25kaWNpb25WZW50YT48TWVkaW9QYWdvPjAzPC9NZWRpb1BhZ28+PERldGFsbGVTZXJ2aWNpbz48TGluZWFEZXRhbGxlPjxOdW1lcm9MaW5lYT4xPC9OdW1lcm9MaW5lYT48Q29kaWdvPjcxMTMzMDEwMDAwMDA8L0NvZGlnbz48Q2FudGlkYWQ+MTwvQ2FudGlkYWQ+PFVuaWRhZE1lZGlkYT5VbmlkPC9VbmlkYWRNZWRpZGE+PERldGFsbGU+ZGV0YWxsZTwvRGV0YWxsZT48UHJlY2lvVW5pdGFyaW8+MTA8L1ByZWNpb1VuaXRhcmlvPjxNb250b1RvdGFsPjEwPC9Nb250b1RvdGFsPjxTdWJUb3RhbD4xMDwvU3ViVG90YWw+PEltcHVlc3RvPjxDb2RpZ28+MDE8L0NvZGlnbz48Q29kaWdvVGFyaWZhPjA4PC9Db2RpZ29UYXJpZmE+PFRhcmlmYT4xMzwvVGFyaWZhPjxNb250bz4xLjM8L01vbnRvPjwvSW1wdWVzdG8+PE1vbnRvVG90YWxMaW5lYT4xMS4zPC9Nb250b1RvdGFsTGluZWE+PC9MaW5lYURldGFsbGU+PC9EZXRhbGxlU2VydmljaW8+PFJlc3VtZW5GYWN0dXJhPjxDb2RpZ29UaXBvTW9uZWRhPjxDb2RpZ29Nb25lZGE+Q1JDPC9Db2RpZ29Nb25lZGE+PFRpcG9DYW1iaW8+NTg1LjY5PC9UaXBvQ2FtYmlvPjwvQ29kaWdvVGlwb01vbmVkYT48VG90YWxTZXJ2R3JhdmFkb3M+MDwvVG90YWxTZXJ2R3JhdmFkb3M+PFRvdGFsU2VydkV4ZW50b3M+MDwvVG90YWxTZXJ2RXhlbnRvcz48VG90YWxNZXJjYW5jaWFzR3JhdmFkYXM+MTA8L1RvdGFsTWVyY2FuY2lhc0dyYXZhZGFzPjxUb3RhbE1lcmNhbmNpYXNFeGVudGFzPjA8L1RvdGFsTWVyY2FuY2lhc0V4ZW50YXM+PFRvdGFsR3JhdmFkbz4xMDwvVG90YWxHcmF2YWRvPjxUb3RhbEV4ZW50bz4wPC9Ub3RhbEV4ZW50bz48VG90YWxFeG9uZXJhZG8+MDwvVG90YWxFeG9uZXJhZG8+PFRvdGFsVmVudGE+MTA8L1RvdGFsVmVudGE+PFRvdGFsRGVzY3VlbnRvcz4wPC9Ub3RhbERlc2N1ZW50b3M+PFRvdGFsVmVudGFOZXRhPjEwPC9Ub3RhbFZlbnRhTmV0YT48VG90YWxJbXB1ZXN0bz4xLjM8L1RvdGFsSW1wdWVzdG8+PFRvdGFsQ29tcHJvYmFudGU+MTEuMzwvVG90YWxDb21wcm9iYW50ZT48L1Jlc3VtZW5GYWN0dXJhPjwvRmFjdHVyYUVsZWN0cm9uaWNhPg=="`;
